test/colors/conversions.jl

```diff
@@ -0,0 +1,155 @@
+            if cs ∈ cspaces_4
+                i = rand(ds4...)
+                @test size(colorify(cs, i)) == (ds4[1:end - 2]..., ds4[end])
+            elseif cs ∈ (AGray, GrayA)
+                i = rand(ds2...)
+                @test size(colorify(cs, i)) == (ds4[1:end - 2]..., ds4[end])
+            elseif cs ∈ (Gray,)
+                i = rand(ds1...)
+                @test size(colorify(cs, i)) == (ds4[1:end - 2]..., ds4[end])
+            else
+                i = rand(ds3...)
+                @test size(colorify(cs, i)) == (ds4[1:end - 2]..., ds4[end])
+            end
+        end
+    end
+    @testset "Testing reversibilty" begin
+        for cs in cspaces_all
+            if cs ∈ cspaces_4
+                i = rand(ds4...)
+                @test channelify(colorify(cs, i)) == i
+            elseif cs ∈ (AGray, GrayA)
+                i = rand(ds2...)
+                @test channelify(colorify(cs, i)) == i
+            elseif cs ∈ (Gray,)
+                i = rand(ds1...)
+                @test channelify(colorify(cs, i)) == i
+            else
+                i = rand(ds3...)
+                @test channelify(colorify(cs, i)) == i
+            end
+        end
+    end
+    @testset "Testing Differentiability" begin
+        # channelview
+        @testset "Testing channelview" begin
+            for cs in cspaces_with_random
+                i = rand(cs, ds...)
+                if cs ∈ (Gray,)
+                    @test channelview(gradient(x -> sum(channelview(x)), i)[1]) == ones(Float64, ds...)
+                elseif cs ∈ (AGray, GrayA)
+                    @test channelview(gradient(x -> sum(channelview(x)), i)[1]) == ones(Float64, 2, ds...)
+                elseif cs ∈ (BGRA, ABGR, RGBA, ARGB)
+                    @test channelview(gradient(x -> sum(channelview(x)), i)[1]) == ones(Float64, 4, ds...)
+                else
+                    @test channelview(gradient(x -> sum(channelview(x)), i)[1]) == ones(Float64, 3, ds...)
+                end
+            end
+        end
+        # colorview
+        # colorview(T, gray1, gray2...) tests will have to be written.
+        @testset "Testing colorview" begin
+            for cs in cspaces_all
+                if cs ∈ cspaces_4
+                    i = rand(ds4[end - 1], ds4[1:end - 2]..., ds4[end])
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
+                elseif cs ∈ (AGray, GrayA)
+                    i = rand(ds2[end - 1], ds2[1:end - 2]..., ds2[end])
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
+                elseif cs ∈ (Gray,)
+                    i = rand(ds1[1:end - 2]..., ds1[end])
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
+                else
+                    i = rand(ds3[end - 1], ds3[1:end - 2]..., ds3[end])
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
```

Comment:
A cleaner design might be to use the if else tree to get `i` or the args and testing gradient only once.


Also you should add tests for differentiability of colorify, channelify and have separate tests for channelview etc as well

Reply:
I already had tests for `channelview` and `colorview`. Do I need to write the `channelify` and `colorify` differentiability tests?

Reply:
I just mean that testing them as individual kernels and verifying they do the right thing is good practice in unit testing.